### PR TITLE
Add quicklook mapping features

### PR DESCRIPTION
**Pull request by nekhtiari**
This PR provides the implementation of quicklooks map via folium. An example of final result would be like this:
Screenshot 2020-09-08 at 15 44 31

### CODE REVIEW

File: requirements-dev.txt [hunk 1]
```diff
@@ -1,6 +1,7 @@
 mkdocs
 mkdocs-material
-pandoc
+#pandoc==2.0.a4
+pypandoc
```
Comment (rodrigoalmeida94) [inline requirements-dev.txt:4]: I guess this you can remove entirely

File: tests/test_utils.py [hunk 2]
```diff
@@ -274,3 +277,50 @@ def test_download_result_from_gcs():
...
+        / "mock_data/quicklooks/quicklook_f8c03432-cec1-41b7-a203-4d871a03290f.jpg"
+    )
+    filepaths = [quicklook_1, quicklook_2]
+
+    m = _map_images(plot_file_format, gdf, filepaths)
+    m._repr_html_()
+    out = m._parent.render()
+
+    assert re.search("Image 1 - f8c03432-cec1-41b7-a203-4d871a03290f", out)
+    assert re.search("Image 2 - 16e18e15-c941-4aae-97cd-d67b18dc9f6e", out)
```
Comment (rodrigoalmeida94) [inline tests/test_utils.py:305]: Nice :)

File: up42/job.py [hunk 3]
```diff
@@ -283,71 +258,21 @@ def _highlight_function(feature):  # pylint: disable=unused-argument
...
+        m = _map_images(
+            plot_file_format=plot_file_format,
+            result_df=df,
+            filepaths=self.results,
+            aoi=None,
+            show_images=show_images,
+            show_features=True,
+            name_column=name_column,
+            save_html=save_html,
+        )
```
Comment (rodrigoalmeida94) [inline up42/job.py:273]: Beautifull!

File: up42/job.py [hunk 4]
```diff
@@ -3,33 +3,21 @@
...
     download_results_from_gcs_without_unpacking,
 )
 
-try:
-    from IPython.display import display
-    from IPython import get_ipython
-except ImportError:
-    # No Ipython installed, Installed but run in shell
-    pass
-
```
Comment (rodrigoalmeida94) [inline up42/job.py:32]: I guess this is ok to remove because we don't use the display any more correct?

File: up42/utils.py [hunk 5]
```diff
@@ -300,6 +320,116 @@ def _plot_images(
...
+def _map_images(
+    plot_file_format: List[str],
+    result_df: GeoDataFrame,
+    filepaths,
+    aoi=None,
+    show_images=True,
+    show_features=False,
+    name_column: str = "id",
+    save_html: Path = None,
+):
```
Comment (rodrigoalmeida94) [inline up42/utils.py:332]: can you add a return type? I guess will be folium.Map?

File: up42/utils.py [hunk 6]
```diff
@@ -300,6 +320,116 @@ def _plot_images(
...
+    # Collapse layer control with too many features.
+    collapsed = bool(result_df.shape[0] > 4)
+    folium.LayerControl(position="bottomleft", collapsed=collapsed).add_to(m)
+
+    if save_html:
+        if not Path(save_html).exists():
+            Path(save_html).mkdir(parents=True, exist_ok=True)
+        filepath = Path(save_html) / "final_map.html"
+        with filepath.open("w") as f:
+            f.write(m._repr_html_())
```
Comment (rodrigoalmeida94) [inline up42/utils.py:429]: Nice!

**Review by rodrigoalmeida94 — COMMENTED** [review]
Love it! How's the code coverage? I supposed it must have increased. Just an idea: you can attach a picture of the output/or the html file to the PR so everyone can see the result.

File: up42/utils.py [hunk 7]
```diff
@@ -181,6 +183,24 @@ def folium_base_map(
...
+
+def _highlight_function(feature):  # pylint: disable=unused-argument
+    return {
+        "fillColor": "#ffaf00",
+        "color": "red",
+        "weight": 3.5,
+        "dashArray": "5, 5",
+    }
+
+
```
Comment (rodrigoalmeida94) [inline up42/utils.py:203]: Can we try to get rid of the pylint disable things?

File: requirements-dev.txt (hunk 1 above)
Comment (nekhtiari) [reply requirements-dev.txt:4]: Right, I will remove!

File: up42/job.py (hunk 4 above)
Comment (nekhtiari) [reply up42/job.py:32]: Right, for your info @chrieke

File: up42/utils.py (hunk 5 above)
Comment (nekhtiari) [reply up42/utils.py:332]: Nice, I did not think about it, I will add it then.

File: up42/utils.py (hunk 7 above)
Comment (nekhtiari) [reply up42/utils.py:203]: Will try

**Review by markusuwe — APPROVED** [review]
LGTM!
Very quick implementation too.

File: up42/utils.py [hunk 8]
```diff
@@ -300,6 +320,116 @@ def _plot_images(
...
+        logger.error(
+            f"The length of the imgaes {len(raster_filepaths)} is not equal to number of"
+            f"row in scenes {len(result_df)}"
+        )
+        raise
+
+    try:
+        feature_names = result_df[name_column].to_list()
+    except KeyError:
+        feature_names = ""
```
Comment (marooned20) [inline up42/utils.py:370]: Here `try/except/else` construct would be more pythonic. So essentially `feature_names = ""` would be part of the else block.

File: up42/utils.py [hunk 9]
```diff
@@ -300,6 +320,116 @@ def _plot_images(
...
+            style_function=_style_function,
+            highlight_function=_highlight_function,
+        ).add_to(m)
+
+    if show_features:
+        for idx, row in result_df.iterrows():  # type: ignore
+            try:
+                feature_name = row.loc[name_column]
+            except KeyError:
+                feature_name = ""
```
Comment (marooned20) [inline up42/utils.py:385]: here too, same as above comment

File: up42/utils.py [hunk 10]
```diff
@@ -300,6 +320,116 @@ def _plot_images(
...
+        for idx, (raster_fp, feature_name) in enumerate(
+            zip(raster_filepaths, feature_names)
+        ):
+            with rasterio.open(raster_fp) as src:
+                if src.meta["crs"] is None:
+                    dst_array = src.read()[:3, :, :]
+                    minx, miny, maxx, maxy = list_bounds[idx]
+                else:
+                    # Folium requires 4326, streaming blocks are 3857
+                    with WarpedVRT(src, crs="EPSG:4326") as vrt:
```
Comment (marooned20) [inline up42/utils.py:408]: Nice! And thought through. `WarpedVRT` to go!

**Review by rodrigoalmeida94 — APPROVED** [review]

File: up42/utils.py (hunk 8 above)
Comment (nekhtiari) [reply up42/utils.py]: Ok as we discussed I will ignore this!

File: up42/job.py [hunk 11]
```diff
@@ -253,35 +242,22 @@ def upload_results_to_bucket(
...
+    def map_results(
+        self, show_images: bool = True, name_column: str = "uid", save_html=None
+    ) -> folium.Map:
         """
         Displays data.json, and if available, one or multiple results geotiffs.
 
         Args:
             show_images: Shows images if True (default), only features if False.
             name_column: Name of the feature property that provides the Feature/Layer name.
+            save_html: The path for saving folium map as html file.
```
Comment (chrieke) [inline up42/job.py:254]: "With default `None` no file is saved" etc. should be mentioned.

File: up42/tools.py [hunk 12]
```diff
@@ -71,6 +78,7 @@ def read_vector_file(
 
         if df.crs.to_string() != "EPSG:4326":
             df = df.to_crs(epsg=4326)
+        df.geometry = df.geometry.buffer(0)
```
Comment (chrieke) [inline up42/tools.py:81]: hm this was removed in a recent PR, can you check if this was reintroduced by accident? Seems you are on top of master.

File: up42/tools.py [hunk 13]
```diff
@@ -223,6 +231,44 @@ def plot_quicklooks(
...
+        filepaths: List = None,
+        name_column: str = "id",
+        save_html: Path = None,
+    ) -> folium.Map:
+        """
+        Plots the downloaded quicklooks (filepaths saved to self.quicklooks of the
+        respective object, e.g. job, catalog).
+
+        Args:
+            scenes: GeoDataFrame of scenes, results of catalog.search()
```
Comment (chrieke) [inline up42/tools.py:247]: aoi & save_html missing in dosctring

File: up42/utils.py [hunk 14]
```diff
@@ -300,6 +320,119 @@ def _plot_images(
...
+    raster_filepaths = [
+        path for path in filepaths if Path(path).suffix in plot_file_format
+    ]
+
+    # Make sure the quicklooks images are equal to the number of features in the scenes
+    try:
+        assert len(list_bounds) == len(raster_filepaths)
+    except AssertionError:
+        logger.error(
+            f"The length of the imgaes {len(raster_filepaths)} is not equal to number of"
```
Comment (chrieke) [inline up42/utils.py:365]:
```suggestion
            f"The number of downloaded quicklooks {len(raster_filepaths)} is not equal to the number of catalog search results {result_df.shape[0]}"
```

Comment (chrieke) [reply up42/utils.py:365]: {len(result_df)} gives the number of columns, `df.shape[0]` or 'len(result_df.index)` is the rows. Is a bit weird in geopandas.

File: up42/utils.py [hunk 15]
```diff
@@ -300,6 +320,119 @@ def _plot_images(
...
+            f"row in scenes {len(result_df)}"
+        )
+        raise
```
Comment (chrieke) [inline up42/utils.py:368]: Logger message is really nice, but not sure this should be raised, we can still display something even if its not all of them no? Did that come up as an issue that not all quicklooks were downloaded or just for precaution?

File: up42/utils.py [hunk 16]
```diff
@@ -300,6 +320,119 @@ def _plot_images(
...
+                    name=f"Image {idx + 1} - {feature_name}",
+                )
+            )
+
+    # Collapse layer control with too many features.
+    collapsed = bool(result_df.shape[0] > 4)
+    folium.LayerControl(position="bottomleft", collapsed=collapsed).add_to(m)
+
+    if save_html:
+        if not Path(save_html).exists():
```
Comment (chrieke) [inline up42/utils.py:428]: `save_html = Path(save_html)` once at the top probably best.

File: up42/utils.py [hunk 17]
```diff
@@ -300,6 +320,119 @@ def _plot_images(
...
+                    bounds=[[miny, minx], [maxy, maxx]],  # different order.
+                    name=f"Image {idx + 1} - {feature_name}",
+                )
+            )
+
+    # Collapse layer control with too many features.
+    collapsed = bool(result_df.shape[0] > 4)
+    folium.LayerControl(position="bottomleft", collapsed=collapsed).add_to(m)
+
+    if save_html:
```
Comment (chrieke) [inline up42/utils.py:427]: nice!

File: up42/utils.py [hunk 18]
```diff
@@ -374,6 +507,7 @@ def any_vector_to_fc(
             if df.crs.to_string() != "EPSG:4326":
                 df = df.to_crs(epsg=4326)
 
+    df.geometry = df.geometry.buffer(0)
```
Comment (chrieke) [inline up42/utils.py:510]: same here with the recent pr.

**Review by chrieke — COMMENTED** [review]
Really cool! Screenshot up top looks ace! A few minor comments then LGTM

File: up42/utils.py (hunk 16 above)
Comment (nekhtiari) [reply up42/utils.py:428]: Ok, I will do

File: up42/utils.py (hunk 18 above)
Comment (nekhtiari) [reply up42/utils.py:510]: Removed!

File: up42/utils.py (hunk 15 above)
Comment (nekhtiari) [reply up42/utils.py:368]: It was more for precaution, I can also remove it? what do you think?

File: up42/tools.py (hunk 13 above)
Comment (nekhtiari) [reply up42/tools.py:247]: Good catch, added!

File: up42/tools.py (hunk 12 above)
Comment (nekhtiari) [reply up42/tools.py:81]: Removed!

**Review by thaisbendixen42 — APPROVED** [review]
Super cool Nikoo! LGTM

**Pull request merged**